tls/policy/s2n_policy_writer.c

```diff
@@ -0,0 +1,187 @@
+        RESULT_GUARD_POSIX(s2n_stuffer_printf(stuffer, "certificate keys:\n"));
+        for (size_t i = 0; i < policy->certificate_key_preferences->count; i++) {
+            RESULT_GUARD_POSIX(s2n_stuffer_printf(stuffer, "- %s\n",
+                    policy->certificate_key_preferences->certificate_keys[i]->name));
+        }
+    }
+
+    if (policy->kem_preferences && policy->kem_preferences != &kem_preferences_null) {
+        RESULT_GUARD_POSIX(s2n_stuffer_printf(stuffer, "pq:\n"));
+        RESULT_GUARD_POSIX(s2n_stuffer_printf(stuffer, "- revision: %i\n",
+                policy->kem_preferences->tls13_pq_hybrid_draft_revision));
+
+        if (policy->kem_preferences->kem_count > 0) {
+            RESULT_GUARD_POSIX(s2n_stuffer_printf(stuffer, "- kems:\n"));
+            for (size_t i = 0; i < policy->kem_preferences->kem_count; i++) {
+                RESULT_GUARD_POSIX(s2n_stuffer_printf(stuffer, "-- %s\n",
+                        policy->kem_preferences->kems[i]->name));
+            }
+        }
+
+        RESULT_GUARD_POSIX(s2n_stuffer_printf(stuffer, "- kem groups:\n"));
+        for (size_t i = 0; i < policy->kem_preferences->tls13_kem_group_count; i++) {
+            RESULT_GUARD_POSIX(s2n_stuffer_printf(stuffer, "-- %s\n",
+                    policy->kem_preferences->tls13_kem_groups[i]->name));
+        }
+    }
+
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT s2n_security_policy_write_to_stuffer(const struct s2n_security_policy *policy,
+        s2n_policy_format format, struct s2n_stuffer *stuffer)
+{
+    RESULT_ENSURE_REF(policy);
+    RESULT_ENSURE_REF(stuffer);
+
+    switch (format) {
+        case S2N_POLICY_FORMAT_DEBUG_V1:
+            RESULT_GUARD(s2n_security_policy_write_format_v1_to_stuffer(policy, stuffer));
+            break;
+        default:
+            RESULT_BAIL(S2N_ERR_INVALID_ARGUMENT);
+    }
+
+    return S2N_RESULT_OK;
+}
+
+int s2n_security_policy_write_length(const struct s2n_security_policy *policy,
+        s2n_policy_format format, uint32_t *length)
+{
+    POSIX_ENSURE_REF(policy);
+    POSIX_ENSURE_REF(length);
+
+    DEFER_CLEANUP(struct s2n_stuffer stuffer = { 0 }, s2n_stuffer_free);
+    POSIX_GUARD(s2n_stuffer_growable_alloc(&stuffer, 1024));
+
+    POSIX_GUARD_RESULT(s2n_security_policy_write_to_stuffer(policy, format, &stuffer));
+
+    *length = s2n_stuffer_data_available(&stuffer);
+
+    return S2N_SUCCESS;
+}
+
+int s2n_security_policy_write_bytes(const struct s2n_security_policy *policy,
+        s2n_policy_format format, uint8_t *buffer, uint32_t buffer_length)
+{
+    POSIX_ENSURE_REF(policy);
+    POSIX_ENSURE_REF(buffer);
+
+    uint32_t required_size = 0;
+    POSIX_GUARD(s2n_security_policy_write_length(policy, format, &required_size));
+    POSIX_ENSURE(buffer_length >= required_size, S2N_ERR_INSUFFICIENT_MEM_SIZE);
+
+    DEFER_CLEANUP(struct s2n_stuffer stuffer = { 0 }, s2n_stuffer_free);
+    POSIX_GUARD(s2n_stuffer_growable_alloc(&stuffer, required_size));
```

Comment:
Is there a reason to write to an intermediate buffer here? Couldn't we create a stuffer from the user's buffer and write directly to that?